### ISSUE TITLE
Feature Request: Refactor `inc` to be of type `ArrayHolder` instead of `IncludeDirsHolder`

### ISSUE THREAD

**Issue by mitchgrout** [mesonbuild/meson#12012]
`inc` objects are backed by the custom `IncludeDirsHolder` type, which internally stores a list of items. This contrasts to `files` which uses the more generic `ArrayHolder` with the stored element type being `File`. Because of this difference, a `files` objects can be iterated and appended to, while an `inc` object cannot be interacted with in any way. 

Since `inc` does not export any public methods, the conversion to `ArrayHolder` should not break any existing configurations. The conversion would also support using custom linters which need access to the project include directories

This request is derived from the need to invoke a custom lint tool, which needs access to: 

1. The source file to be linted, and
2. The list of include paths used to resolve header paths

**Example**

```meson
project('testing', 'c')

sources  = files('foo.c', 'bar.c', 'baz.c')
foreach source: sources
  message('source = @0@'.format(source))
endforeach

includes = include_directories('inc/foo', 'inc/bar', 'inc/baz')
foreach include: includes
  message('include = @0@'.format(include))
endforeach
```

**Desired behavior:**

```text
$ meson setup build
...
Message: source = foo.c
Message: source = bar.c
Message: source = baz.c
Message: include = inc\foo
Message: include = inc\bar
Message: include = inc\baz
...
```

**Current behaviour:** (`<= 1.2.0`)

```text
$ meson setup build
...
Message: source = foo.c
Message: source = bar.c
Message: source = baz.c

meson.build:9:17: ERROR: Items of foreach loop do not support iterating
...
```

**Comment by mitchgrout** [comment]
In retrospect the change required to support my use case would be a bit more than just making `inc` iterable. The more fleshed-out example would be something to the effect of:

```meson
# Assume these are assembled via repeated `subdir()` calls
sources  = files('foo.c', 'bar.c', 'baz.c')
includes = include_directories('inc')

iwyu    = find_program('include-what-you-use')
targets = []
command = [iwyu, '@INPUT@']
foreach include: includes # would need iteration, but also 
  command += '-I@0@'.format(include) # would need formatting relative to _some_ directory
endforeach

foreach source: sources
  targets += custom_target(input: source, output: '@PLAINNAME@.iwyu', capture: true, command: command)
endforeach
alias_target('iwyu', targets)
```

As noted, #9486 could be a suitable approach, but it would put more burden on situations like mine due to needing to filter out other assorted compiler bits. It could definitely work in my use case though.

An alternative could be to re-scope this issue to instead export a method on `inc` which can produce a `list[str]` of the _absolute_ include paths, via `build.IncludeDirs.to_string_list`, which would then adjust my code fragment to

```meson
command = [iwyu, '@INPUT@']
foreach dir: includes.absolute_paths() # -> list[str]
  command += '-I@0@'.format(dir)
endforeach
```

I think there is still some room to implement this issue as proposed, if not for my use case directly; this feels like a rather natural thing to try, given the apparent symmetry of the two methods:

```meson
sources  = files()
includes = include_directories()

sources  += 'foo.c' # OK
includes += 'inc'   # Error
```

**Comment by eli-schwartz** [comment]
> An alternative could be to re-scope this issue to instead export a method on `inc` which can produce a `list[str]` of the _absolute_ include paths, via `build.IncludeDirs.to_string_list`, which would then adjust my code fragment to

Indeed, that could be a valid path forward -- one less focused on the mechanics of "make `inc` support iteration". :)

> As noted, #9486 could be a suitable approach, but it would put more burden on situations like mine due to needing to filter out other assorted compiler bits. It could definitely work in my use case though.

Not necessarily -- a `dep` can already do `dep.partial_dependency(includes: true)` -- but wrapping an `inc` in a `declare_dependency()` to get a dependency-to-array-of-strings is a layer of indirection that could be argued to be... less than perfectly elegant. :)

**Comment by mitchgrout** [comment]
I'll take a crack at implementing a method like `inc.absolute_paths() -> list[str]` and see if it fits my need. I take it the preference would be to create a new issue scoped for that change, and then close this one out?

I think there's also a value in trying to implement #9486 although I'm not sure if anyone's looking into that; what's the meson stance on having the same behaviour implemented in multiple ways? Would we prefer one-or-the-other, or happy to have them side by side?

**Comment by bonzini** [comment]
Closing as the PRs add a new method without making the object an array.

**Comment by mitchgrout** [comment]
> Closing as the PRs add a new method without making the object an array.

Apologies to post on a closed issue, but no PR was accepted that resolves this.
In retrospect, the solution to my problem is not to allow `inc` objects to be iterated, but instead to have some introspection into the build-plan, i.e. the same contents as `compile-commands.json` or similar.
